### PR TITLE
Refactor/csl777

### PR DESCRIPTION
**Pull request by neongreen**

### CODE REVIEW

File: core/Pos/Util/HVect.hs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+       ) where
+
+import           Universum        hiding (empty)
+
+import           Data.Vector      (Vector)
+import qualified Data.Vector      as V
+import           GHC.TypeLits     (type (+))
+import           Unsafe.Coerce    (unsafeCoerce)
+
+-- | A heterogeneous vector. @HVect [Int, Bool, Char]@ is a vector of with
```
Comment (gromakovsky) [inline core/Pos/Util/HVect.hs]: > of with three elements

File: core/Pos/Util/Context.hs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+instance MonadContext m => MonadContext (ReaderT s m) where
+    type ContextType (ReaderT s m) = ContextType m
+instance MonadContext m => MonadContext (StateT s m) where
+    type ContextType (StateT s m) = ContextType m
+instance MonadContext m => MonadContext (ExceptT s m) where
+    type ContextType (ExceptT s m) = ContextType m
+
+{- |
+Using 'getFullContext' if not very convenient because usually we don't need
```
Comment (gromakovsky) [inline core/Pos/Util/Context.hs]: s/if/is?

File: core/Pos/Util/HVect.hs (hunk 1 above)
Comment (neongreen) [inline core/Pos/Util/HVect.hs]: ✓

File: core/Pos/Util/Context.hs (hunk 2 above)
Comment (neongreen) [inline core/Pos/Util/Context.hs]: ✓

File: core/Pos/Util/Context.hs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+       ) where
+
+import           Universum
+
+import           Control.Lens   (Getting)
+import           Pos.Util.HVect (HVect)
+import qualified Pos.Util.HVect as HVect
+
+class ExtractContext a s where
+    extractContext :: s -> a
```
Comment (ChShersh) [inline core/Pos/Util/Context.hs]: Why there is no `| s -> a` here? Can some example be given?

File: core/Pos/Util/Context.hs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+class Monad m => MonadContext m where
+    type ContextType m
+    getFullContext :: m (ContextType m)
+
+{- |
+Using 'getFullContext' if not very convenient because usually we don't need
```
Comment (ChShersh) [inline core/Pos/Util/Context.hs]: > if not very convenient
Typo

File: src/Pos/Launcher/Runner.hs [hunk 5]
```diff
@@ -358,13 +360,15 @@ runCH params@NodeParams {..} sscNodeContext act = do
             { ncJLFile = jlFile
             , ncSscContext = sscNodeContext
             , ncBlkSemaphore = semaphore
-            , ncLrcSync = lrcSync
+            , ncLrcContext = LrcContext
```
Comment (ChShersh) [inline src/Pos/Launcher/Runner.hs]: Seems like `RecordWildCards` can be used here to simplify this constructor into `let ctx = NodeContext{..}`

File: src/Pos/Lrc/Context.hs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+data LrcContext = LrcContext
+    {
+    -- | Primitive for synchronization with LRC.
+      lcLrcSync :: !(TVar LrcSyncData)
+    }
+
+-- | Data used for LRC syncronization. First value is __False__ iff
+-- LRC is running now. Second value is last epoch for which we have
+-- already computed LRC.
+type LrcSyncData = (Bool, EpochIndex)
```
Comment (ChShersh) [inline src/Pos/Lrc/Context.hs]: Custom data type would be better :(

File: src/Pos/Wallet/WalletMode.hs [hunk 7]
```diff
@@ -333,7 +334,8 @@ instance MonadIO m => MonadUpdates (WalletDB m) where
 -- | Instance for full node
 instance (Ssc ssc, MonadIO m, WithLogger m) =>
          MonadUpdates (PC.ContextHolder ssc m) where
-    waitForUpdate = liftIO . takeMVar . PC.ncUpdateSemaphore =<< PC.getNodeContext
+    waitForUpdate = liftIO . takeMVar =<<
```
Comment (ChShersh) [inline src/Pos/Wallet/WalletMode.hs]: `liftIO` is not needed here

File: src/node/Main.hs [hunk 8]
```diff
@@ -242,7 +244,7 @@ updateTriggerWorker
     => ([WorkerSpec (RawRealMode ssc)], OutSpecs)
 updateTriggerWorker = first pure $ worker mempty $ \_ -> do
     logInfo "Update trigger worker is locked"
-    void $ liftIO . takeMVar . ncUpdateSemaphore =<< getNodeContext
+    void $ liftIO . takeMVar =<< askContext ucUpdateSemaphore
```
Comment (ChShersh) [inline src/node/Main.hs]: And here (liftIO)

File: src/Pos/Launcher/Runner.hs (hunk 5 above)
Comment (neongreen) [inline src/Pos/Launcher/Runner.hs]: ✓

File: src/Pos/Lrc/Context.hs (hunk 6 above)
Comment (neongreen) [inline src/Pos/Lrc/Context.hs]: not my code and too lazy to fix :(

File: src/Pos/Wallet/WalletMode.hs (hunk 7 above)
Comment (neongreen) [inline src/Pos/Wallet/WalletMode.hs]: will fix as part of `hlint` pass #2

File: src/node/Main.hs (hunk 8 above)
Comment (neongreen) [inline src/node/Main.hs]: and this too

File: src/Pos/Lrc/Context.hs (hunk 6 above)
Comment (neongreen) [inline src/Pos/Lrc/Context.hs]: okay, I'll fix it

File: src/Pos/Lrc/Context.hs [hunk 9]
```diff
@@ -16,4 +16,7 @@ data LrcContext = LrcContext
 -- | Data used for LRC syncronization. First value is __False__ iff
 -- LRC is running now. Second value is last epoch for which we have
 -- already computed LRC.
-type LrcSyncData = (Bool, EpochIndex)
+data LrcSyncData = LrcSyncData
+    { lrcNotRunning    :: !Bool
+    , lastEpochWithLrc :: !EpochIndex
```
Comment (gromakovsky) [inline src/Pos/Lrc/Context.hs]: Name is not conventional 😒 
Dunno if it's important though.

**Pull request merged**